akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -300,6 +410,231 @@ abstract class GraphStageLogic private[stream] (inCount: Int, outCount: Int) {
   final protected def isClosed[T](out: Outlet[T]): Boolean = (interpreter.portStates(conn(out)) & OutClosed) != 0
 
   /**
+   * Read a number of elements from the given inlet and continue with the given function,
+   * suspending execution if necessary. This action replaces the [[InHandler]]
+   * for the given inlet if suspension is needed and reinstalls the current
+   * handler upon receiving the last `onPush()` signal (before invoking the `andThen` function).
+   */
+  final protected def readN[T](in: Inlet[T], n: Int)(andThen: Array[T] ⇒ Unit): Unit =
+    if (n < 0) throw new IllegalArgumentException("cannot read negative number of elements")
+    else if (n == 0) andThen(new Array[Any](0).asInstanceOf[Array[T]])
```

Comment:
In fact: yes, this would be necessary if actually going with Arrays because otherwise there could be CCEs (which is why there is currently a ClassTag missing as well). In order to sidestep these issues I changed it to an ArrayBuffer.